src/thread_event.c

```diff
@@ -0,0 +1,109 @@
+	 * the next allocation call).
+	 */
+	*tsd_thread_allocated_thresholdp_get(tsd) = threshold;
+
+	if (likely(threshold <= THREAD_ALLOCATED_THRESHOLD_FAST_MAX)) {
+		*tsd_thread_allocated_threshold_fastp_get(tsd) = threshold;
+	} else {
+		/*
+		 * Set thread_allocated_threshold_fast to zero so that
+		 * thread_event_fast() will always trigger.
+		 */
+		*tsd_thread_allocated_threshold_fastp_get(tsd) = 0;
+	}
+}
+
+void
+thread_event_trigger(tsd_t *tsd, uint64_t usize, bool delay_event) {
+	/* usize has already been added to thread_allocated. */
+	uint64_t thread_allocated_after = tsd_thread_allocated_get(tsd);
+
+	/* The subtraction is intentionally susceptible to underflow. */
+	uint64_t accumbytes = thread_allocated_after -
+	    tsd_thread_allocated_last_event_get(tsd);
+
+	if (accumbytes < usize) {
+		/* accumbytes overflowed. */
+		accumbytes = UINT64_MAX;
+	}
+
+	/*
+	 * In case none of the events has its condition on, the UINT64_MAX
+	 * initialization will automatically delay the next event as much as
+	 * possible.
+	 */
+	uint64_t wait = UINT64_MAX;
+
+#define W(event, condition)						\
+	if (condition) {						\
+		uint64_t event_wait =					\
+		    tsd_##event##_event_wait_get(tsd);			\
+		assert(event_wait > 0);					\
+		if (event_wait <= accumbytes) {				\
+			event_wait = 0;					\
+		} else {						\
+			event_wait -= accumbytes;			\
+		}							\
+		if (unlikely(delay_event || !tsd_nominal(tsd) ||	\
+		    tsd_reentrancy_level_get(tsd) > 0) &&		\
+		    event_wait == 0) {					\
+			/* Delay the event. */				\
+			event_wait = 1;					\
+		}							\
+		*tsd_##event##_event_waitp_get(tsd) = event_wait;	\
+		/*							\
+		 * If there is a single event, then the remaining wait	\
+		 * time won't be positive, and we rely on a		\
+		 * thread_event_update() call later to properly set the	\
+		 * next threshold; if there are multiple events, then	\
+		 * here we can get the minimum remaining wait time to	\
+		 * the next already set event.				\
+		 */							\
+		if (event_wait > 0 && event_wait < wait) {		\
+			wait = event_wait;				\
+		}							\
+	}
+
+	ITERATE_OVER_ALL_WAITS
+#undef W
+
+	*tsd_thread_allocated_last_eventp_get(tsd) = thread_allocated_after;
+	thread_event_adjust_thresholds_helper(tsd, thread_allocated_after,
+	    wait);
+}
+
+void
+thread_event_update(tsd_t *tsd) {
+	assert(tsd_nominal(tsd));
+	assert(tsd_reentrancy_level_get(tsd) == 0);
+
+	uint64_t wait = thread_event_min_wait_helper(tsd);
+	uint64_t last_event = tsd_thread_allocated_last_event_get(tsd);
+	thread_event_adjust_thresholds_helper(tsd, last_event, wait);
+
+	/* The subtraction is intentionally susceptible to underflow. */
+	uint64_t overlooked = tsd_thread_allocated_get(tsd) - last_event;
+
+	/* The subtraction is intentionally susceptible to underflow. */
+	if (unlikely(overlooked >= tsd_thread_allocated_threshold_get(tsd) -
+	    last_event)) {
+		thread_event_trigger(tsd, overlooked, true);
```

Comment:
Been digging into this for a while. I think this might be caused by on ralloc / xalllocx path, the discrepancy between the usize passed to thread_event() and the actual allocation usize. Since we always increment thread.allocated at the start, we have some fixup to do here. If we adjust properly, we should be able to ensure this never happens.

Reply:
The `rallocx()` / `xallocx()` path can be a cause, but from what I observed, a non-zero `overlooked` does not seem to be caused by the wrong allocation size, but instead, it seems the `prof_sample_threshold_update()` function can be called in some circumstances even no prof sampling event is triggered, and in such cases `overlooked` will not be zero, and in some cases may have surpassed the new threshold. I also saw other cases before where `thread_allocated` was incremented by interrupting `malloc()` calls (i.e. those with `reentrancy_level > 0`) that take place after `thread_event()` but *before* `thread_event_update()`.

Meanwhile, I think we should have some adjustment mechanism, even just for the correctness of `thread_allocated` on the `rallocx()` / `xallocx()` path. How about what I proposed earlier: having the previous `last_event` and `threshold` backed up in `thread_event()` and a new `thread_event_rollback()` function?

Reply:
Yes I agree that we need some mechanism to adjust / rollback. Let's talk more in person.